### PR TITLE
(TWILL-118) Added support for Distributed Lock for Twill application

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

File: twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/ReentrantDistributedLock.java [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.Nullable;
+
+/**
+ * A reentrant distributed lock implementation that uses ZooKeeper. It uses the receipt described in
```
Comment (albertshau) [inline twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/ReentrantDistributedLock.java]: typo, receipt -> recipe

Comment (ghelmling) [inline twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/ReentrantDistributedLock.java]: "receipt" -> "recipe"

**Comment by albertshau** [comment]
:+1:

File: twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/ReentrantDistributedLock.java (hunk 1 above)
Comment (chtyim) [inline twill-zookeeper/src/main/java/org/apache/twill/internal/zookeeper/ReentrantDistributedLock.java]: fixed

**Comment by chtyim** [comment]
Merged to master. Close PR.

**Pull request closed without merging**